sdks/java/core/src/main/java/org/apache/beam/sdk/io/WholeFileIO.java

```diff
@@ -0,0 +1,241 @@
+      );
+
+      String filePattern = getFilePattern().get();
+
+      PCollection<String> filePatternPCollection = input.apply(Create.of(filePattern));
+
+      PCollection<ResourceId> resourceIds = filePatternPCollection.apply(
+          ParDo.of(
+              new DoFn<String, ResourceId>() {
+                @ProcessElement
+                public void processElement(ProcessContext c) {
+                  String filePattern = c.element();
+                  try {
+                    List<MatchResult> matchResults = FileSystems.match(
+                        Collections.singletonList(filePattern));
+                    for (MatchResult matchResult : matchResults) {
+                      List<MatchResult.Metadata> metadataList = matchResult.metadata();
+                      for (MatchResult.Metadata metadata : metadataList) {
+                        ResourceId resourceId = metadata.resourceId();
+                        c.output(resourceId);
+                      }
+                    }
+                  } catch (IOException e) {
+                    e.printStackTrace();
+                  }
+                }
+              }
+          )
+      );
+
+      PCollection<KV<String, byte[]>> files = resourceIds.apply(
+          ParDo.of(
+              new DoFn<ResourceId, KV<String, byte[]>>() {
+                @ProcessElement
+                public void processElement(ProcessContext c) {
+                  ResourceId resourceId = c.element();
+
+                  try {
+                    ReadableByteChannel channel = FileSystems.open(resourceId);
+                    ByteBuffer byteBuffer = ByteBuffer.allocate(8192);
+                    ByteString byteString = ByteString.EMPTY;
+
+                    while (channel.read(byteBuffer) != -1) {
+                      byteBuffer.flip();
+                      byteString = byteString.concat(ByteString.copyFrom(byteBuffer));
+                      byteBuffer.clear();
+                    }
+
+                    KV<String, byte[]> kv = KV.of(
+                        resourceId.getFilename(),
+                        byteString.toByteArray()
+                    );
+
+                    c.output(kv);
+                  } catch (IOException e) {
+                    e.printStackTrace();
+                  }
+                }
+              }
+          )
+      );
+
+      return files;
+    }
+  }
+
+  /**
+   * Implements write().
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<KV<String, byte[]>>, PDone> {
+
+    @Nullable abstract ValueProvider<ResourceId> getOutputDir();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setOutputDir(ValueProvider<ResourceId> outputDir);
+
+      abstract Write build();
+    }
+
+    public Write to(String outputDir) {
+      return to(FileBasedSink.convertToFileResourceIfPossible(outputDir));
```

Comment:
Seems like this one is explicitly intended to be a directory, but the function you call here tries to make it a regular file and only converts it to a directory if it can't. Just use FileSystems.matchNewResource(..., true)